### PR TITLE
Eng 7157

### PR DESCRIPTION
**Pull request by hhakimi53**
Logic update to embedded jetty server to add directives for cache-control and send them over the wire in compressed form (gzip). The resources that are stamped with cache-control directives also supply etag directive so that the file gets retransmitted only when it gets changed. These resources are stylesheets (resides in css folder), images (images folder) and javascript (js folder). JSON and other requests to server are not stamped with caching or gzipped.

Changes are under test.

### CODE REVIEW

File: src/frontend/org/voltdb/utils/HTTPAdminListener.java [hunk 1]
```diff
@@ -328,6 +351,41 @@ public void handle(String target,
...
+    /*
+     * Utility handler class to enable caching of static resources.
+     * The static resources are package in jar file
+     */
+    class CacheStaticResourceHandler extends ResourceHandler {
+        // target Directory location for folder w.r.t. resource base folder - dbmonitor
+        public CacheStaticResourceHandler(final String target) {
+            super();
+            final String path = VoltDB.class.getResource(RESOURCE_BASE + File.separator + target).toExternalForm();
+            m_log.debug("Resource base path: " + path);
```
Comment (akhanzode) [inline src/frontend/org/voltdb/utils/HTTPAdminListener.java]: Put this under isDebugenabled()

File: src/frontend/org/voltdb/utils/HTTPAdminListener.java [hunk 2]
```diff
@@ -328,6 +351,41 @@ public void handle(String target,
...
+            setResourceBase(path);
+            setCacheControl("max-age=600, private");
```
Comment (akhanzode) [inline src/frontend/org/voltdb/utils/HTTPAdminListener.java]: Get age from hidden property HTTP_STATIC_CACHE_MAXAGE and put it in wiki page.
Make default 24h

File: src/frontend/org/voltdb/utils/HTTPAdminListener.java [hunk 3]
```diff
@@ -328,6 +351,41 @@ public void handle(String target,
...
+            setEtags(true);
```
Comment (akhanzode) [inline src/frontend/org/voltdb/utils/HTTPAdminListener.java]: Add comment why combination of Etag and max age is needed.

File: src/frontend/org/voltdb/utils/HTTPAdminListener.java [hunk 4]
```diff
@@ -328,6 +351,41 @@ public void handle(String target,
...
+
+        @Override
+        public void handle(String target,
+                           Request baseRequest,
+                           HttpServletRequest request,
+                           HttpServletResponse response)
+                           throws IOException, ServletException {
+            super.handle(target, baseRequest, request, response);
+            if (!baseRequest.isHandled()) {
+                m_log.debug("Failed to process static resource: " + Paths.get(getResourceBase()));
```
Comment (akhanzode) [inline src/frontend/org/voltdb/utils/HTTPAdminListener.java]: Guard by isDebug

File: src/frontend/org/voltdb/utils/HTTPAdminListener.java [hunk 5]
```diff
@@ -1060,17 +1118,40 @@ public HTTPAdminListener(
...
+                    dbMonitorHandler,
+                    cssResourceHandler,
+                    imageResourceHandler,
+                    jsResourceHandler
+                    });
+
+            GzipHandler compressResourcesHandler = new GzipHandler();
+            compressResourcesHandler.setHandler(handlers);
+
+            compressResourcesHandler.addExcludedMimeTypes(JSON_CONTENT_TYPE);
```
Comment (akhanzode) [inline src/frontend/org/voltdb/utils/HTTPAdminListener.java]: Why not compress the JSON payload?

Comment (hhakimi53) [inline src/frontend/org/voltdb/utils/HTTPAdminListener.java]: In initial implementation JSON payloads were also compressed though with that the browser was using a lot of cpu cycles, every json payload has to be decompressed on browser end (there is similar penalty on server side for compressing it to). After turning compression off for json payloads, the cpu utilization by browser went back to normal state. So, have put the json payload in exclude list

File: src/frontend/org/voltdb/utils/HTTPAdminListener.java (hunk 2 above)
Comment (hhakimi53) [inline src/frontend/org/voltdb/utils/HTTPAdminListener.java]: sure

File: src/frontend/org/voltdb/utils/HTTPAdminListener.java (hunk 3 above)
Comment (hhakimi53) [inline src/frontend/org/voltdb/utils/HTTPAdminListener.java]: sure

**Comment by akhanzode** [comment]
Looks good. squash merge if jenkins is happy.

**Pull request merged**